### PR TITLE
GH-428: Cannot assign undefined in a typed destructuring pattern #428

### PR DESCRIPTION
**Pull request by qtran-n4**
Task: https://github.com/eclipse/n4js/issues/428

### CODE REVIEW

**Comment by qtran-n4** [comment]
TODOs to discuss with @mor-n4:

- Quick mode in PolyProcessor_ObjectLiteral: 
```
val quickMode = !haveUsableExpectedType && !TypeUtils.isInferenceVariable(expectedTypeRef)
```
changed to
```
val quickMode = !haveUsableExpectedType && !TypeUtils.isInferenceVariable(expectedTypeRef) && !expectedTypeRef.topType;
```

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend [hunk 1]
```diff
@@ -175,6 +199,73 @@ package class PolyProcessor extends AbstractPolyProcessor {
 		infCtx.solve;
 	}
 
+	/** Calculate expected type of a destructure pattern based on its structure */
+	private def TypeRef calculateExpectedType(DestructNode destructNode, RuleEnvironment G) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend]: move this and next method to DestructureHelper?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend [hunk 2]
```diff
@@ -154,12 +163,27 @@ package class PolyProcessor extends AbstractPolyProcessor {
 		// (until the expectedType judgment is integrated into AST traversal, we have to invoke this judgment here;
 		// in case of not-well-behaving expectedType rules, we use 'null' as expected type, i.e. no expectation)
 		// TODO integrate expectedType judgment into AST traversal and remove #isProblematicCaseOfExpectedType()
-		val expectedTypeRef = if (!rootPoly.isProblematicCaseOfExpectedType) {
+		var expectedTypeRef = if (!rootPoly.isProblematicCaseOfExpectedType) {
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend [hunk 3]
```diff
@@ -154,12 +163,27 @@ package class PolyProcessor extends AbstractPolyProcessor {
...
+		else {
+			null
+		};
+
+		if (rootDestructNode !== null) {
+			expectedTypeRef = calculateExpectedType(rootDestructNode, G)
+		}
 		// call #processExpr() (this will recursively call #processExpr() on nested expressions, even if non-poly)
-		val typeRef = processExpr(G, rootPoly, expectedTypeRef, infCtx, cache);
+		var typeRef = processExpr(G, rootPoly, expectedTypeRef, infCtx, cache);
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor_ObjectLiteral.xtend [hunk 4]
```diff
@@ -78,8 +78,8 @@ package class PolyProcessor_ObjectLiteral extends AbstractPolyProcessor {
 		// quick mode as a performance tweak:
 		val haveUsableExpectedType = expectedTypeRef !== null
 				&& (expectedTypeRef.useSiteStructuralTyping || expectedTypeRef.defSiteStructuralTyping); // FIXME reconsider
-		val quickMode = !haveUsableExpectedType && !TypeUtils.isInferenceVariable(expectedTypeRef);
 
+		val quickMode = !haveUsableExpectedType && !TypeUtils.isInferenceVariable(expectedTypeRef);
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor_ObjectLiteral.xtend]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend [hunk 5]
```diff
@@ -154,12 +163,27 @@ package class PolyProcessor extends AbstractPolyProcessor {
...
+		var rootDestructNode = if (rootPoly.eContainer instanceof VariableBinding) {
+			DestructNode.unify(rootPoly.eContainer as VariableBinding)
+		} else if (rootPoly.eContainer instanceof AssignmentExpression) {
+			DestructNode.unify(rootPoly.eContainer as AssignmentExpression)
+		} else if (rootPoly.eContainer instanceof ForStatement) {
+			DestructNode.unify(rootPoly.eContainer as ForStatement)
+		}
+		else {
+			null
+		};
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend]: move this also to a helper method `calculateExpectedType(Expression)` (in DestructureHelper)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend (hunk 1 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend]: done.

**Comment by qtran-n4** [comment]
Thanks!

**Pull request merged**